docs/kepler_model_server/pipeline.md

```diff
@@ -1,25 +1,54 @@
+As shown in the below figure, a power model can be attributed by the path it was trained. Training pipeline is an abstract of power model training that applies a set of learning methods to a different combination of energy source labels, available metrics, and idle/control plane power isolation choices to each specific group of nodes/machines. 
 
-## Model Output Type
+A training pipeline starts from reading the Kepler-exporting metrics from Prometheus query (prom) and finally submits an archived models to the model database (model DB). 
 
-**AbsPower:** the predicted power is a node power that includes nearly-static power at idling state and dynamic power when running the process on the machine.
+From Kepler queries, the default extractor generates a dataframe with the following columns.
+
+timestamp|features|labels[unit]\_[#unit]\_[component]\_power|node type
+---|---|---|---
+timestamp|e.g.,<br>cgroupfs_cpu_usage_us<br>cgroupfs_memory_usage_bytes<br>cgroupfs_system_cpu_usage_us<br>cgroupfs_user_cpu_usage_us|e.g.,<br>package_0_package_power<br>package_1_package_power<br>package_0_core_power<br>package_1_core_power<br>package_0_uncore_power<br>package_1_uncore_power<br>package_0_dram_power<br>package_1_dram_power|node_type
+
+
+![](../fig/model-server-e2e.png)
+
+<!-- TOC tocDepth:2..3 chapterDepth:2..6 -->
+
+- [Labeling energy source](#labeling-energy-source)
+- [Idle power/Control plane power](#idle-powercontrol-plane-power)
+- [Available metrics](#available-metrics)
+- [Learning methods](#learning-methods)
+- [Node Spec](#node-spec)
+
+<!-- /TOC -->
 
-**DynPower:** the predicted power is per-process/per-container power that excludes nearly-static power at idling state.
+## Labeling energy source
 
-### Power source
+`energy source` or `source` refers to the source (power meter) that provides an energy number. Each source provides one or more `energy components`. Currently supported source are shown as below.
 
-Power source|Power components
+Energy/power source|Energy/power components
 ---|---
 [rapl](../design/kepler-energy-sources.md#rapl---running-average-power-limit)|package, core, uncore, dram
 [acpi](../design/kepler-energy-sources.md#using-kernel-driver-xgene-hwmon)|platform
+||
 
-## Model Features
-The features are groupped by the metric sources.
+## Idle power/Control plane power
 
-### Feature group
+`isolate` is a mechanism to separate the power portion that is consumed on the node in an idle state or the power portion that is consumed by operating systems and control plane processes. These portions of power is more than zero even if the metric utilization of workload is zero. We called the models those are trained after isolating these power portions as `DynPower` models. At the same time, the models those are trained without isolation are called `AbsPower` models. `DynPower` model is used to estimate `container power` and `process power` while `AbsPower` model is used to estimate `node power`.
```

Comment:
> [...] the power portion that is consumed by operating systems and control plane processes. These portions of power is more than zero even if the metric utilization of workload is zero.

What do you mean about that?

The idle power is higher than 0 if the utilization of workload is zero, but the dynamic power should be 0 if the utilization of workload is zero!

Although we might not have the OS CPU utilization via cgroup or kubelet metrics, the eBPF metrics will provide the OS and control-plane CPU utilization which must match the resource utilization with the power utilization.

Additionally, we should maybe have a better distinction between isolating idle power and isolating other workload dynamic power (i.e., removing the control plane).

Reply:
I would like to say about the part that we isolated from the AbsPower (AbsPower - DynPower).
Yes, I agree that we should elaborate more about the isolation. Because in the control plane case, it mean we also remove utilization from the control plane process too (done by trainIsolator).  Please feel free to modify the doc.